vm/systemSmartContracts/governance.go

```diff
@@ -385,972 +394,573 @@ func (g *governanceContract) delegateVote(args *vmcommon.ContractCallInput) vmco
 
-// saveGeneralProposal saves a proposal into the storage
-func (g *governanceContract) saveGeneralProposal(reference []byte, generalProposal *GeneralProposal) error {
-	marshaledData, err := g.marshalizer.Marshal(generalProposal)
-	if err != nil {
-		return err
+	if totalVotes.Cmp(minQuorumOutOfStake) == -1 {
+		g.eei.AddReturnMessage("Proposal did not reach minQuorum")
+		proposal.Passed = false
+		return nil
 	}
-	key := append([]byte(proposalPrefix), reference...)
-	g.eei.SetStorage(key, marshaledData)
-
-	return nil
-}
 
-func (g *governanceContract) getEndNonceForProposal(reference []byte) uint64 {
-	proposal, err := g.getGeneralProposal(reference)
-	if err == vm.ErrProposalNotFound {
-		return 0
+	minVetoOfTotalVotes := core.GetIntTrimmedPercentageOfValue(totalVotes, float64(baseConfig.MinVetoThreshold))
+	if proposal.Veto.Cmp(minVetoOfTotalVotes) >= 0 {
+		proposal.Passed = false
+		g.eei.AddReturnMessage("Proposal vetoed")
+		return nil
 	}
 
-	if err != nil {
-		return math.MaxUint64
+	minPassOfTotalVotes := core.GetIntTrimmedPercentageOfValue(totalVotes, float64(baseConfig.MinPassThreshold))
+	if proposal.Yes.Cmp(minPassOfTotalVotes) >= 0 && proposal.Yes.Cmp(proposal.No) > 0 {
+		g.eei.AddReturnMessage("Proposal passed")
+		proposal.Passed = true
+		return nil
 	}
 
-	return proposal.EndVoteNonce
+	g.eei.AddReturnMessage("Proposal rejected")
+	proposal.Passed = false
+	return nil
 }
 
-// getGeneralProposal returns a proposal from storage
-func (g *governanceContract) getGeneralProposal(reference []byte) (*GeneralProposal, error) {
-	key := append([]byte(proposalPrefix), reference...)
-	marshaledData := g.eei.GetStorage(key)
-
+func (g *governanceContract) getActiveFundForDelegator(delegationAddress []byte, address []byte) (*big.Int, error) {
+	marshaledData := g.eei.GetStorageFromAddress(delegationAddress, address)
 	if len(marshaledData) == 0 {
-		return nil, vm.ErrProposalNotFound
+		return big.NewInt(0), nil
 	}
 
-	generalProposal := &GeneralProposal{}
-	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
+	dData := &DelegatorData{}
+	err := g.marshalizer.Unmarshal(dData, marshaledData)
 	if err != nil {
 		return nil, err
 	}
 
-	return generalProposal, nil
-}
-
-// proposalExists returns true if a proposal already exists
-func (g *governanceContract) proposalExists(reference []byte) bool {
-	key := append([]byte(proposalPrefix), reference...)
-	marshaledData := g.eei.GetStorage(key)
-	return len(marshaledData) > 0
-}
+	if len(dData.ActiveFund) == 0 {
+		return big.NewInt(0), nil
+	}
 
-// getValidProposal returns a proposal from storage if it exists, or it is still valid/in-progress
-func (g *governanceContract) getValidProposal(reference []byte) (*GeneralProposal, error) {
-	proposal, err := g.getGeneralProposal(reference)
+	marshaledData = g.eei.GetStorageFromAddress(delegationAddress, dData.ActiveFund)
+	activeFund := &Fund{Value: big.NewInt(0)}
```

Comment:
useless set here, it will be automatically rewritten by the Unmarshal call. To be 100% we do not return nil from this function, we need an if before the return on L784

Reply:
done